Python/bytecodes.c

```diff
@@ -790,10 +790,15 @@ dummy_func(
             // Deopt unless 0 <= sub < PyList_Size(list)
             DEOPT_IF(!_PyLong_IsNonNegativeCompact((PyLongObject *)sub));
             Py_ssize_t index = ((PyLongObject*)sub)->long_value.ob_digit[0];
-            DEOPT_IF(index >= PyList_GET_SIZE(list));
+            DEOPT_IF(!LOCK_OBJECT(list));
+            if (index >= PyList_GET_SIZE(list)) {
+                UNLOCK_OBJECT(list);
+                DEOPT_IF(true);
+            }
             STAT_INC(BINARY_SUBSCR, hit);
             PyObject *res_o = PyList_GET_ITEM(list, index);
             assert(res_o != NULL);
+            UNLOCK_OBJECT(list);
             Py_INCREF(res_o);
```

Comment:
I think we need to incref `res_o` before we unlock the list. Otherwise, it's possible that we could push an object that was never stored at `index` if the following happened:

1. We unlock `list`.
2. Another thread stores a new object in `list` at the same index. `list` held the last reference to `res_o` and the assignment destroys the last reference, freeing the memory pointed to by `res_o`. 
3. The memory that`res_o` points to is reused by another thread for a new object.
4. We incref `res_o`, but now it's a completely different object that was never stored in the list.